From b2b40eb53cd3027139a31a0a6947ca2152d0fa79 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 16 Sep 2024 15:25:52 +0200
Subject: [PATCH 1238/1695] drivers: stm32_gpio: check pin secure state of
 pinctrl states

Allow pinctrl states to cover non-secure pins if the pin description
matches the secure state of the pin.

This change removes the warning trace message when getting a pinctrl
state with pins not matching their current secure state. This warning
will only be emitted when the pin is effectively used (when the pinctrl
state is applied with stm32_pinctrl_conf_apply()).

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 636e80e529c7cae415f2f1223dd22fd07e3c6e96 drivers: stm32_gpio: check pin secure state of pinctrl states]
Change-Id: I0e9abb1e673304b086eeca34e643a0d3c918f767
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/404686
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_gpio.c | 25 ++++++++++++++++---------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/core/drivers/stm32_gpio.c b/core/drivers/stm32_gpio.c
index 3ccf29da3..1c680151e 100644
--- a/core/drivers/stm32_gpio.c
+++ b/core/drivers/stm32_gpio.c
@@ -15,6 +15,7 @@
 #include <drivers/stm32_gpio.h>
 #include <drivers/stm32_rif.h>
 #include <dt-bindings/gpio/stm32mp_gpio.h>
+#include <dt-bindings/pinctrl/stm32-pinfunc.h>
 #include <io.h>
 #include <keep.h>
 #include <kernel/dt.h>
@@ -128,6 +129,7 @@
  * @pupd: One of GPIO_PUPD_*
  * @od: One of GPIO_OD_*
  * @af: Alternate function numerical ID between 0 and 15
+ * @nsec: Hint on expected secure state of the pin: 0 if secure, 1 otherwise
  */
 struct gpio_cfg {
 	uint16_t mode:		2;
@@ -136,6 +138,7 @@ struct gpio_cfg {
 	uint16_t pupd:		2;
 	uint16_t od:		1;
 	uint16_t af:		4;
+	uint16_t nsec:		1;
 };
 
 /*
@@ -586,6 +589,7 @@ static int get_pinctrl_from_fdt(const void *fdt, int node,
 		uint32_t alternate = 0;
 		uint32_t odata = 0;
 		bool opendrain = false;
+		bool pin_non_secure = true;
 
 		pincfg = fdt32_to_cpu(*cuint);
 		cuint++;
@@ -596,6 +600,8 @@ static int get_pinctrl_from_fdt(const void *fdt, int node,
 
 		mode = pincfg & DT_GPIO_MODE_MASK;
 
+		pin_non_secure = pincfg & STM32_PIN_NSEC;
+
 		switch (mode) {
 		case 0:
 			mode = GPIO_MODE_INPUT;
@@ -656,6 +662,7 @@ static int get_pinctrl_from_fdt(const void *fdt, int node,
 			ref->cfg.pupd = pull;
 			ref->cfg.od = odata;
 			ref->cfg.af = alternate;
+			ref->cfg.nsec = pin_non_secure;
 
 			bank_ref = stm32_gpio_get_bank(bank);
 
@@ -669,10 +676,6 @@ static int get_pinctrl_from_fdt(const void *fdt, int node,
 				     fdt_get_name(fdt, consumer_node, NULL),
 				     bank + 'A', pin);
 				do_panic = true;
-			} else if (!pin_is_secure(bank_ref, pin)) {
-				IMSG("WARNING: node %s requests pinctrl with non-secure pin %c%u, check st,protreg in GPIO bank node",
-				     fdt_get_name(fdt, consumer_node, NULL),
-				     bank + 'A', pin);
 			}
 		}
 
@@ -1302,15 +1305,19 @@ static TEE_Result stm32_pinctrl_conf_apply(struct pinconf *conf)
 	for (n = 0; n < pin_count; n++) {
 		struct stm32_gpio_bank *bank = stm32_gpio_get_bank(p[n].bank);
 
-		if (pin_is_secure(bank, p[n].pin))
+		if (p[n].cfg.nsec == !pin_is_secure(bank, p[n].pin))
 			continue;
 
 		if (IS_ENABLED(CFG_INSECURE)) {
-			IMSG("WARNING: apply pinctrl with non-secure pin %c%u",
-			     p[n].bank + 'A', p[n].pin);
+			IMSG("WARNING: apply pinctrl for %ssecure pin %c%u that is %ssecure",
+			     p[n].cfg.nsec ? "non-" : "",
+			     p[n].bank + 'A', p[n].pin,
+			     pin_is_secure(bank, p[n].pin) ? "" : "non-");
 		} else {
-			EMSG("Apply pinctrl with non-secure pin %c%u",
-			     p[n].bank + 'A', p[n].pin);
+			EMSG("Apply pinctrl for %ssecure pin %c%u that is %ssecure",
+			     p[n].cfg.nsec ? "non-" : "",
+			     p[n].bank + 'A', p[n].pin,
+			     pin_is_secure(bank, p[n].pin) ? "" : "non-");
 			error = true;
 		}
 	}
-- 
2.47.2

